modules/system/classes/FileManifest.php

```diff
@@ -0,0 +1,200 @@
+     * @param string $root The root folder to get the file list from.
+     * @param array $modules An array of modules to include in the file manifest.
+     */
+    public function __construct($root = null, array $modules = null)
+    {
+        if (isset($root)) {
+            $this->setRoot($root);
+        } else {
+            $this->setRoot(base_path());
+        }
+
+        if (isset($modules)) {
+            $this->setModules($modules);
+        } else {
+            $this->setModules(Config::get('cms.loadModules', ['System', 'Backend', 'Cms']));
+        }
+    }
+
+    /**
+     * Sets the root folder.
+     *
+     * @param string $root
+     * @throws ApplicationException If the specified root does not exist.
+     */
+    public function setRoot($root)
+    {
+        if (is_string($root)) {
+            $this->root = realpath($root);
+
+            if ($this->root === false || !is_dir($this->root)) {
+                throw new ApplicationException(
+                    'Invalid root specified for the file manifest.'
+                );
+            }
+        }
+
+        return $this;
+    }
+
+    /**
+     * Sets the modules.
+     *
+     * @param array $modules
+     */
+    public function setModules(array $modules)
+    {
+        $this->modules = array_map(function ($module) {
+            return strtolower($module);
+        }, $modules);
+
+        return $this;
+    }
+
+    /**
+     * Gets a list of files and their corresponding hashsums.
+     *
+     * @return array
+     */
+    public function getFiles()
+    {
+        if (count($this->files)) {
+            return $this->files;
+        }
+
+        $files = [];
+
+        foreach ($this->modules as $module) {
+            $path = $this->root . DIRECTORY_SEPARATOR . 'modules' . DIRECTORY_SEPARATOR . $module;
+
+            if (!is_dir($path)) {
+                continue;
+            }
+
+            foreach ($this->findFiles($path) as $file) {
+                $files[$this->getFilename($file)] = md5_file($file);
+            }
+        }
+
+        return $this->files = $files;
+    }
+
+    /**
+     * Gets the checksum of a specific install.
+     *
+     * @return array
+     */
+    public function getModuleChecksums()
+    {
+        if (!count($this->files)) {
+            $this->getFiles();
+        }
+
+        $modules = [];
+        foreach ($this->modules as $module) {
+            $modules[$module] = '';
+        }
+
+        foreach ($this->files as $path => $hash) {
+            // Determine module
+            $module = explode(DIRECTORY_SEPARATOR, $path)[2];
+
+            $modules[$module] .= $hash;
+        }
+
+        return array_map(function ($moduleSum) {
+            return md5($moduleSum);
+        }, $modules);
+    }
+
+    /**
+     * Finds all files within the path.
+     *
+     * @param string $basePath The base path to look for files within.
+     * @return array
+     */
+    protected function findFiles(string $basePath)
```

Comment:
@LukeTowers doesn't FileDatasource only support a single level of subdirectories?

Reply:
@bennothommo the FileDatasource class directly doesn't care about the nesting level, it's merely the Halcyon Model & Builder that implement the check for nesting. See the example code I pasted in https://github.com/octobercms/october/pull/4615#issuecomment-611696247, it returned all paths.

Reply:
@LukeTowers the [`select`](https://github.com/octobercms/library/blob/develop/src/Halcyon/Datasource/FileDatasource.php#L112) method of the FileDatasource class does only support a single layer though. That method's the only reason I'd use FileDatasource, as it would save a lot of the conditional code I originally had.

Reply:
@bennothommo the selectOne method supports arbitrary depth AFAIK though

Reply:
@LukeTowers I was meaning using the `select` method to get all paths that match a certain extension through the `extensions` option. That would allow me to get rid of [this block of code](https://github.com/octobercms/october/blob/4363d129a7b124a58e91b67446e617606cbbac9f/modules/system/classes/FileManifest.php#L165-L171), at least. `selectOne` would only get a single file.

Reply:
Ah, I see. Why are you only looking at those specific file extensions? Why not include all of them?

Reply:
@LukeTowers mainly to keep the filesize of the manifest down. I figured we'd very unlikely push out a build that just makes changes to the Markdown files, for example.

Reply:
That being said, I probably should include `htm` and `html` files though :P

Reply:
We wouldn't push out a build no, but I personally think it's important to know if there are any changes period. Using markdown as an example, when it comes to the rainlab docs plugin, it could be useful to know if the docs files have been modified locally, etc.

Are you able to test the filesize difference between the filtering you have right now (including htm & html) and not doing any filtering period?

Reply:
@LukeTowers It's ~409kb with the specified extensions (plus `htm` files), and ~496kb with everything. So I suppose it's not a huge drama including them all.

Reply:
Yeah, let's go with including them all then.

Reply:
Done :)